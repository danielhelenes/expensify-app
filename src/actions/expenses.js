//since we don't have any jsx, we don't need to import react.

import uuid from 'uuid';

 export const addExpense = ({description = '', note = '', amount = 0, createdAt = 0} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note, // this is the same of note: note
    amount,
    createdAt
  }
})

/*
// action above is the same as the one below:
export const addExpense = (anything = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description: anything.description,
    note: anything.note, // this is the same of note: note
    amount: anything.amount,
    createdAt: anything.createdAt
  }
})*/


// REMOVE EXPENSES

export const removeExpense = ({id} = {}) => ({ //question: Why do we need the default here?
  type: 'REMOVE_EXPENSE',
  id
})

// EDIT EXPENSES

export const editExpense = (id, updates) => ({ //we don't need defaults here.
  type: 'EDIT_EXPENSE',
  id,
  updates
});
