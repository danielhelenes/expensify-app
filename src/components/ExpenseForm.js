import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


// const date = new Date(); we are not going to use this because the API is super complicated. never use it!(date object in Javascript)


export default class ExpenseForm extends React.Component { //cannot log (?)

  constructor(props) { //this props come from the components that use this component - addexpense and editExpense!!!
    super(props);

    this.state = {
        description: props.expense ? props.expense.description : '',
        note: props.expense ? props.expense.note : '',
        amount: props.expense ? (props.expense.amount /100).toString() : '',
        createdAt: props.expense ? moment(props.expense.createdAt) : moment(), //calling the now moment.
        calendarFocused: false,
        error: ''
    };
  };


  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description })) // this is the same of description: description.
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(()=> ({ note })) // this is the same of description: description.
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) { //regex expression from regex101.com
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) { //if to prevent the user to cler the date value. if there is a createdAt, do that:
      this.setState(() => ({ createdAt }))

    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  };

  onSubmit = (e) => {
    const a = this.props;
    console.log(a);
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      //set error state = please provide description and Amount
      this.setState(() => ({ error: 'Please provide a description and amount' }));
    } else {
      //clear the error.
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10)*100, // 10 means the base we are working on(?) - 100 is because we are working in cents
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render(){
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
            >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
};
