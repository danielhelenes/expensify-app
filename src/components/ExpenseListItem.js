// export a stateless functional component
// description, amount, createdAt
// use map in expenseList.js

import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';


const ExpenseListItem = ({ id, description, amount, createdAt }) => (//props is not defined. console logs each expenses object with amount, createdat, etc... but not expenses in general  (it lists all the objects of items)
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{description}</h3>
      <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
    </div>
    <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
  </Link>
);

export default ExpenseListItem;



// -----------------------------//////////////// -----------------------------//////////////


//Same as above:
// const ExpenseListItem = (props) =>  ( //props is not defined. console logs each expenses object with amount, createdat, etc... but not expenses in general  (it lists all the objects of items)
//   <div>
//     <Link to={`/edit/${props.id}`}>
//       <h3>{props.description}</h3>
//     </Link>
//     <p>{props.amount} - {props.createdAt}</p>
//   </div>
// );

// MY SOLUTION => passing mapStateToProps!

// const ExpenseListItem = (props) => (
//   <div>
//     <h3>{props.description}</h3>
//     <h3>{props.id}</h3>
//     <p>{props.amount} - {props.createdAt}</p>
//     <button onClick={() => props.dispatch(removeExpense({id: props.id}))}>Remove</button>
//   </div>
// );
//
// const mapStateToProps = (state) => {
//   return {
//     expenses: state.expenses
//   }
// };
//
// export default connect(mapStateToProps)(ExpenseListItem);

// -----------------------------//////////////// -----------------------------//////////////
