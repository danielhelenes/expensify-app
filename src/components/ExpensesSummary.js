import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

//expense count = how many expenses are visible
// expenses total =  what's the total of visible expenses

const ExpensesSummary = (props) => (
  <div>
    <h2>Viewing {props.expenses.length} expenses totalling {numeral(props.total / 100).format('$0,0.00')}</h2>
  </div>
);

// numeral(props.total / 100).format('$0,0.00')

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
    total: getExpensesTotal(selectExpenses(state.expenses, state.filters))
  }
}


export default connect(mapStateToProps)(ExpensesSummary);
