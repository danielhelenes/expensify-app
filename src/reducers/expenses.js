// Expenses Reducer - gets all the expenses related components

//question: why does this reducer has access to action if they are in different folders and there is no import here??!?!

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter((expense) => {//this is the same as cur. we map through the cur object that state has. we look for the exact id inside of all the objects that has the same id of the one we are deleting, so we can delete it.
        return expense.id !== action.id;
      });

    case 'EDIT_EXPENSE':
    return state.map((expense) => { //this is the same as cur. we map through the cur object that state has. if the id of each cur object matches the id of the action object (the one we are editing), we return the cur object + the updates.
      if (expense.id === action.id) {
        return {...expense, ...action.updates}
      }else { //for all the objects that have a different id than the one we are editing, we return the same object. nothing changes.
        return expense;
      }
    });

    default:
      return state;
  }
};


///IMPORTANT: the state is simply what you add to it. that is the following object:
/*expense: {
  id: uuid(),
  description,
  note, // this is the same of note: note
  amount,
  createdAt
}*/
