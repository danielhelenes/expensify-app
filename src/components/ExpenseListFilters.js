import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

//we nee to pass onChange because otherwise value will be a static value.
//we have access to dispatch from the component below.
//when we pass the value and onchange to forms inputs and selects, the documentation calls this "controlled input"


export class ExpenseListFilters extends React.Component {
    state = {
      calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
      this.props.setStartDate(startDate);
      this.props.setEndDate(endDate);
      console.log(this.props); // access to dispatch and filters
    };

    onFocusedChange = (calendarFocused) => {
      this.setState(() => ({ calendarFocused }));
      console.log(this.props); // access to dispatch and filters
    };

    onTextChange = (e) => {
      this.props.setTextFilter(e.target.value);
    };

    onSelectChange = (e) => {
      if (e.target.value === 'date') {
        this.props.sortByDate();
      } else if (e.target.value === 'amount'){
        this.props.sortByAmount();
      }
    };

  render() {
    return(
      <div>
        <input
          type="text"
          value={this.props.filters.text} // we changed to a class based component, so we need to use this.props instead of props.
          onChange={this.onTextChange}/>
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSelectChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate = {this.props.filters.startDate}
          endDate = {this.props.filters.endDate}
          onDatesChange = {this.onDatesChange}
          focusedInput = {this.state.calendarFocused}
          onFocusChange = {this.onFocusedChange}
          showClearDates = {true}
          numberOfMonths = {1}
          isOutsideRange = {() => false}
        />

      </div>
    );
  }
}


const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate:  () => dispatch(sortByDate()),
  sortByAmount: () =>  dispatch(sortByAmount())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);




// {/* MY SOLUTION */}
//     {/* <select onChange={(e) => {
//       const value = e.target.value;
//       value === 'date' ? props.dispatch(sortByDate()) : props.dispatch(sortByAmount());
//     }}>
//       <option value="date">Date</option>
//       <option value="amount">Amount</option>
//     </select>
