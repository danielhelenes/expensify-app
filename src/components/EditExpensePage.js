import React from 'react';
import { connect } from 'react-redux'; //connect to send this data to the store.
import ExpenseForm from './ExpenseForm';// ExpenseForm to get data from ExpenseForm and use the props. (dispatch, history, location, match, etc.... and expense!)
import { startEditExpense, startRemoveExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense); //can also use props.math.params.id
    this.props.history.push('/dashboard');
  };
  onRemove = () => {
    this.props.startRemoveExpense({id: this.props.expense.id});
    this.props.history.push('/dashboard');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => { //we do that to find the array that matches the id of props.match.parms.id
      return expense.id === props.match.params.id; //this will return true for when the condition is valid, and then add this to the expense.
    })
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data)) //question - didnt understand why to pass props and data.
})


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
