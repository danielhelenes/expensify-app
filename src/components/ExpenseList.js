import React from 'react';
import { connect } from 'react-redux'; //connects your component to the redux store.
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// Unconnected component

export const ExpenseList = (props) => ( //didn't have console.
// console.log(props); //have access to expenses and dispatch

  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
        props.expenses.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No expenses</span>
          </div>
        ) : (
            props.expenses.map((expense) => {//expense is the current obj you are mapping through.
              return <ExpenseListItem key={expense.id} {...expense} />; // this spread operator is actually composed of amount, description etc... that is actually the props used in ExpenseListItem as arguments. intead of the spread, we could have used id={expense.id} amount={expense.amount}
            })
          )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList); // we need to have this setup ('what information from the store we want to connect')('where we want this connection to go') because of their API. because we setup this connect call, expenseList will have access to the props of the object passed in mapStateToProps. this export default connects mapStateToProps with ExpenseList.
