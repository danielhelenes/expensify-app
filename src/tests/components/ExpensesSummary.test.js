import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import getExpensesTotal from '../../selectors/expenses-total';
import expenses from '../fixtures/expenses';



test('should render 0 expenses and 0 total if empty array', () => {
  const expenses = [];
  const total = getExpensesTotal(expenses);
  const wrapper = shallow(<ExpensesSummary expenses={expenses} total={total}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with 1 expenses', () => {
  const expenses = [{
    description: 'coffee',
    amount: 10
  }]
  const total = getExpensesTotal(expenses);
  const wrapper = shallow(<ExpensesSummary expenses={expenses} total={total}/>);
  expect(wrapper).toMatchSnapshot();
});



test('should render ExpensesSummary with multiple expenses', () => {
  const total = getExpensesTotal(expenses);
  const wrapper = shallow(<ExpensesSummary expenses={expenses} total={total}/>);
  expect(wrapper).toMatchSnapshot();
});



//snapshots
