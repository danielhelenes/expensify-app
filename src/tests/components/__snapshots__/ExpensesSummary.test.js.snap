// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 0 expenses and 0 total if empty array 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span />
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with 1 expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span />
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span />
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
